inst/tests/tests.Rraw

```diff
@@ -15369,7 +15369,7 @@ L = list(1:3, NULL, 4:6)
 test(2058.18, length(L), 3L)
 test(2058.19, as.data.table(L), data.table(V1=1:3, V2=4:6))  # V2 not V3        # no
 DT = data.table(a=1:3, b=c(4,5,6))
-test(2058.20, DT[,b:=list(NULL)], data.table(a=1:3))                            # no
+test(2058.20, DT[,b:=list(NULL)], data.table(a=1:3, b=list(NULL)))              # no
```

Comment:
Hmm, this is a bit surprising to me. I suspect this will cause revdep breakage. Here are some examples:

https://github.com/search?q=lang%3AR%20%2F%3A%3D%5Cs*list%5C(%5Cs*NULL%5Cs*%5C)%2F&type=code

It would help to do a before/after on this PR of various ways to add columns, e.g. combinations of

 - adding 1 column vs. multiple columns
 - deleting 1 column vs. multiple columns
 - adding & deleting columns in the same query
 - operations on 1-row vs multi-row tables
 - operations on list- vs non-list columns

Your examples in the main PR body are good but only cover a small part of the above. It may be that we need to cause some breakage for consistency, but we need to understand completely what's changing, what the recommended alternative is, why we can't fix the issue back-compatibly, etc.

e.g. for this test, IINM there's already a recommended way to add a new list column as a plonk (full-column replacement): `b := .(list(NULL))`. Is it not possible (or just strongly ill-advised?) to allow both codepaths to continue to work:

```r
b := list(NULL) # same as b := NULL, column deletion b/c 'b' is not a list to begin with
b := .(list(NULL)) # overwrite 'b' with an 'empty' list column
```

This is a tricky area because I think there's some inherent ambiguity here with list columns.

Reply:
I'm currently working on some tests to run against this PR and previous behavior to see if there are any unexpected changes, will update here.

> e.g. for this test, IINM there's already a recommended way to add a new list column as a plonk (full-column replacement): b := .(list(NULL)). Is it not possible (or just strongly ill-advised?) to allow both codepaths to continue to work:

I believe this only works if b is a list column, in the test mentioned above this would throw an error:
`'list' object cannot be coerced to type 'double'`

Reply:
Added some tests, except I just realized that the changes don't allow us to _add_ any new columns with `list(NULL)` quite yet. I believe I only addressed the replacement of columns with `list(NULL)`. Is adding a new null list column something we want to be able to do as well?

```R
dt = data.table(a = 1:3)
DT[, b := list(NULL)] # warning, doesn't add the new column

# same as
data.table(a = 1:3, b = list(NULL))
# can be done with
DT[, b := .(list(NULL))]
# or
DT[, b := list(list(NULL))]
```

Reply:
> Is adding a new null list column in standard form something we want to be able to do as well? 

I'm not sure it's possible to do unambiguously. What's most important is that users can do what they need to, and there is a way to add "empty" list columns as you noted: `b := .(list(NULL))`. We'd only worry about continuing to support `b := list(NULL)` if there was a back-compatibility issue, which there's not in this case.

Reply:
SGTM. How do you suppose we proceed with this issue? Are the revdep issues worth making for the sake of consistency? Primary goal for me is to close issues, so I'm open to suggestions 😸

Reply:
> How do you suppose we proceed with this issue?

Let's start with the proposed list of the various ways to add/remove columns for list/non-list types. It can also serve as a piece of documentation to add somewhere (maybe FAQ).

Reply:
** Not done yet, just commenting here as I need to change PCs **
Here's a table of current behavior and proposed changes

<table>
<tr>
<td> Type </td> <td> Old </td> <td> Proposed changes </td>
</tr>
<tr>
<td> Replacement of a list column in a single-row data.table with `list(NULL)` with $ or standard form with :=. </td>
<td>

```r
# Deletes list column
DT = data.table(L = list('A'), i = 1)
DT$L = list(NULL)
# or
DT[, L := list(NULL)]
#        i
#    <num>
# 1:     1
```
</td>
<td> 

```r
# Replaces list column with empty list, consistent with data.frame 
# and consistent with current functional form
DT = data.table(L = list('A'), i = 1)
DT$L = list(NULL)
# or
DT[, L := list(NULL)]
# or
DT[, `:=`(L = list(NULL))] # before AND after PR the same
#        i      L
#    <num> <list>
# 1:     1 [NULL]
```
</td>


</tr>
<tr>
<td> Replacement of a list column in a multi-row data.table with `list(NULL)` with $ or with :=. </td>
<td>

```r
# Replaces list column with empty list
DT = data.table(L = list('A', 'B'), i = 1)
DT$L = list(NULL, NULL)
# or
DT[, L := list(NULL, NULL)]
# or
DT[, `:=`(L = list(NULL))]
#        i      L
#    <num> <list>
# 1:     1 [NULL]
# 2:     1 [NULL]
```
</td>
<td>

```r
# Does the same thing, but removes the need to specify "NULL" for every row
DT = data.table(L = list('A', 'B'), i = 1)
DT$L = list(NULL)
# or
DT[, L := list(NULL)]
# or
DT[, `:=`(L = list(NULL))] # Again, same before and after
#        i      L
#    <num> <list>
# 1:     1 [NULL]
# 2:     1 [NULL]
```
</td>
</tr>
<tr>
<td> Adding an empty list column to a single-row data.table. </td>
<td>

```r
# Use a plonk when necessary
DT = data.table(L = list('A'), i = 1)
DT$D = list(list(NULL))
# or
DT[, D := .(list(NULL))]
# or
DT[, `:=`(D = list(NULL))]
#         L     i      D
#    <list> <num> <list>
# 1:      A     1 [NULL]
```
</td>
<td> No changes here! </td>
</tr>
<tr>

<td> Adding an empty list column to a multi-row data.table. </td>
<td>

```r
# Use a plonk when necessary
DT = data.table(L = list('A', 'B'), i = 1)
DT$D = list(list(NULL))
# or
DT[, D := .(list(NULL))]
# or
DT[, `:=`(D = list(NULL))]
#         L     i      D
#    <list> <num> <list>
# 1:      A     1 [NULL]
# 2:      C     1 [NULL]
```
</td>
<td> Again, no changes here! </td>
</tr>
<tr>
<td> </td>